[W5.11][F09-A2]Ewald Hew

ewaldhew opened the pull request:
WIP.
TODO:
Add more JUnit tests.
Add I/O tests.

ewaldhew committed:
EditCommand: New class

* Add UI messages.

ewaldhew committed:
Parser: Make it accept `edit' command

* parser.Parser: Add new format string for command.
Minor renaming.
(prepareEdit): Add argument parse method.
* commands.EditCommand: Add basic constructor.

ewaldhew committed:
EditCommand: Implement.

* data.AddressBook (editPerson): Removes the old person and adds in the new
one.

ewaldhew committed:
AddressBook::editPerson: Rethrow `PersonNotFoundException' after rollback

ewaldhew committed:
Add JUnit tests

ewaldhew committed:
UserGuide.md: Update

ewaldhew committed:
commands.HelpCommand: Update message

ewaldhew committed:
Update I/O tests

ewaldhew committed:
expected.txt: Space/Tab fix

jonathanwj commented on src/seedu/addressbook/commands/EditCommand.java:
```diff
@@ -0,0 +1,72 @@
+        } catch (IndexOutOfBoundsException ie) {
+            return new CommandResult(Messages.MESSAGE_INVALID_PERSON_DISPLAYED_INDEX);
+        } catch (DuplicatePersonException dpe) {
+            return new CommandResult(MESSAGE_DUPLICATE_PERSON);
```
Just a question, why not put MESSAGE_DUPLICATE_PERSON string in the Messages class as well?

ewaldhew replied:
Not exactly the same message, should they be? hmm

kokonguyen191 replied:
If you use MESSAGE_DUPLICATE_PERSON in another **Command**, then I think putting it in **Messages** class is a good idea, else you can just put it here.

jonathanwj reviewed (commented):
Code looks good to me.

kokonguyen191 commented on src/seedu/addressbook/data/AddressBook.java:
```diff
@@ -83,6 +83,29 @@ public void addPerson(Person toAdd) throws DuplicatePersonException {
+     */
+    public void editPerson(ReadOnlyPerson target, Person replacement)
+            throws DuplicatePersonException, PersonNotFoundException {
+        addPerson(replacement);
```
Is this the wanted behavior? **AddressBook** throws **DuplicatePersonException** if it finds a person with the same name and phone number in the list. Now what if I want to change John Doe's email address from _johnd@gmail.com_ to _johnd@u.nus.edu_? Then **AddressBook** will throw **DuplicatePersonException**. Should you add first then remove or should you remove first then add?

ewaldhew replied:
Hmm, I didn't check that duplicate considers name and phone only... intention was to disallow editing a person to be exactly same as another. I'll look into using the other compare function instead.

kokonguyen191 commented on src/seedu/addressbook/commands/EditCommand.java:
```diff
@@ -0,0 +1,72 @@
+import seedu.addressbook.data.person.UniquePersonList.DuplicatePersonException;
+import seedu.addressbook.data.person.UniquePersonList.PersonNotFoundException;
+
+
```
Header comments

ewaldhew committed:
Parser, EditCommand: Add more JUnit tests

kokonguyen191 commented on src/seedu/addressbook/commands/EditCommand.java:
```diff
@@ -0,0 +1,72 @@
+import seedu.addressbook.data.person.UniquePersonList.PersonNotFoundException;
+
+
+public class EditCommand extends Command {
```
Header comment

ewaldhew committed:
EditCommand: Add header comment.

ewaldhew committed:
EditCommandTest: Remove unused stuff.

ewaldhew committed:
AddressBook: Change `editPerson' logic.

`editPerson' would throw a `DuplicatePersonException' if the edited person
only changed their email or address without changing phone number, due to
the check done by `isSamePerson'. Inverting the order of operations fixes
this.

* AddressBook (editPerson): Fix logic.
* EditCommandTest: Update tests.

k-l-a commented on doc/UserGuide.md:
```diff
@@ -66,12 +66,29 @@ Examples:
+  `edit 2 p/98765432 e/johnd@gmail.com a/John street, block 123, #01-01`
+  Edits the 2nd person in the address book.
+* `find Betsy`<br>
+  `edit 1 pp/1234567 e/betsycrowe@gmail.com pa/Newgate Prison t/criminal t/friend`
```
Typo in `pp`?

k-l-a commented on src/seedu/addressbook/commands/EditCommand.java:
```diff
@@ -0,0 +1,74 @@
+     *
+     * @throws IllegalValueException if any of the raw values are invalid
+     */
+    public EditCommand(int targetVisibleIndex,
```
Are all fields mandatory? Maybe you can find a way to only change select fields.

k-l-a commented on src/seedu/addressbook/commands/EditCommand.java:
```diff
@@ -0,0 +1,74 @@
+
+    @Override
+    public CommandResult execute() {
+        try {
```
Try catch block should as close to the line where possible exception happens as possible.

k-l-a commented on test/java/seedu/addressbook/commands/EditCommandTest.java:
```diff
@@ -0,0 +1,287 @@
+import seedu.addressbook.ui.TextUi;
+import seedu.addressbook.util.TestUtil;
+
+public class EditCommandTest {
```
Good job creating JUnit tests for newly added method/classes.

k-l-a reviewed (commented):
Good attempt! Some comments added. Please close PR after reading.

To note, please give a brief summary of what you have enhanced in your PR description. It helps understanding what enhancements you have (or intended to) made.
